### DISCUSSION TITLE
Roadmap Stage 2: Code-aware experiments

### DISCUSSION THREAD

**Discussion by shaunc** [FactFiber/kedro-dvc#7] [Ideas]
### Stage 2: Code-aware experiments

Use dvc experiment tracking to keep track of different versions of a project. Use kedro hooks with [trace](https://docs.python.org/3/library/trace.html) to optionally skip steps.

* [@mzip2 example skip hook](https://gist.github.com/mzjp2/076bfd73b0215bda01ee71186966389d) but without trace.

Pipelines will have to be run in a number of different modes:

* Run everything, snapshot code dependencies
* Run as required, using code dependencies stored and/or dvc versions
* Run without trace for better performance

The intention is to use dvc to manage experiment versions. This may
require additional coupling between kedro runner and dvc; however,
by default we can use [dvclive](https://github.com/iterative/dvclive)
where possible to encapsulate dvc.

**Comment by antonymilne** [comment]
Since you're using it on this repo, I wonder how pytest-watch works and whether some mechanism like that might be useful here 🤔  I don't know how capable it is of tracing at the granularity required (as opposed to just changed files) but at a glance it seems to use [watchdog](https://pypi.org/project/watchdog/) under the hood.

**Reply by shaunc** [reply]
Something like @watchdog is a good idea!
